benchmark/benchmark_driver.py

```diff
@@ -0,0 +1,190 @@
+        return stopping
+
+    benchmarking.addBoth(tear_down)
+
+    return benchmarking
+
+
+def record_samples(
+    samples, version, metric_name, measurement_name, out=stdout
+):
+    """
+    Create a record from the samples and write it.
+
+    :param samples: Sampling results.
+    :param version: Flocker server version.
+    :param metric_name: Sampled operation.
+    :param measurement_name: Samples measurement.
+    :param out: Output file for record.
+    """
+    timestamp = datetime.now().isoformat()
+    artifact = dict(
+        client=dict(
+            flocker_version=flocker_client_version,
+            date=timestamp,
+            working_directory=getcwd(),
+            username=environ[b"USER"],
+            nodename=node(),
+            platform=platform(),
+        ),
+        server=dict(
+            flocker_version=version,
+        ),
+        measurement=measurement_name,
+        metric_name=metric_name,
+        samples=samples,
+    )
+    json.dump(artifact, out)
+
+
+class FastConvergingFakeFlockerClient(
+    proxyForInterface(IFlockerAPIV1Client)
+):
+    def create_dataset(self, *a, **kw):
+        result = self.original.create_dataset(*a, **kw)
+        self.original.synchronize_state()
+        return result
+
+    def move_dataset(self, *a, **kw):
+        result = self.original.move_dataset(*a, **kw)
+        self.original.synchronize_state()
+        return result
+
+    def delete_dataset(self, *a, **kw):
+        result = self.original.delete_dataset(*a, **kw)
+        self.original.synchronize_state()
+        return result
+
+    def create_container(self, *a, **kw):
+        result = self.original.create_container(*a, **kw)
+        self.original.synchronize_state()
+        return result
+
+    def delete_container(self, *a, **kw):
+        result = self.original.delete_container(*a, **kw)
+        self.original.synchronize_state()
+        return result
+
+
+def driver(reactor, control_service_address=None, cert_directory=b"certs",
+           metric_name=b"read-request", measurement_name=b"wallclock",
+           scenario_name=b'ten_ro_req_sec'):
+
+    if control_service_address:
+        cert_directory = FilePath(cert_directory)
+        client = FlockerClient(
+            reactor,
+            host=control_service_address,
+            port=4523,
+            ca_cluster_path=cert_directory.child(b"cluster.crt"),
+            cert_path=cert_directory.child(b"user.crt"),
+            key_path=cert_directory.child(b"user.key"),
+        )
+    else:
+        client = FastConvergingFakeFlockerClient(FakeFlockerClient())
+
+    metric = get_operation(client=client, name=metric_name)
+    measurement = get_measurement(
+        clock=reactor, client=client, name=measurement_name,
+    )
+    scenario = get_scenario(
+        clock=reactor, client=client, name=scenario_name,
+    )
+    version = client.version()
+
+    d = gather_deferreds((metric, measurement, scenario, version))
```

Comment:
:grey_question: Looks like `driver` takes human-inputtable values, turns them into domain objects and then actually starts driving, I'd consider factoring out a separate function that takes the actual domain objects, and having `driver` call that.